src/profile/helpers/editData.ts

```diff
@@ -0,0 +1,202 @@
+import _ from 'lodash';
+
+import {
+  ProfileRoot,
+  ProfileData,
+  AddressNode,
+  EmailNode,
+  PhoneNode,
+  PrimaryAddress,
+  PrimaryEmail,
+  PrimaryPhone,
+  Language,
+} from '../../graphql/typings';
+import { formConstants } from '../constants/formConstants';
+import getAddressesFromNode from '../helpers/getAddressesFromNode';
+import getEmailsFromNode from '../helpers/getEmailsFromNode';
+import getPhonesFromNode from '../helpers/getPhonesFromNode';
+
+export type BasicDataValue = Pick<
+  ProfileData,
+  'firstName' | 'nickname' | 'lastName'
+>;
+export type AdditionalInformationValue = {
+  language: Language;
+};
+export type AddressValue = Pick<
+  AddressNode,
+  'address' | 'city' | 'postalCode' | 'countryCode'
+>;
+
+export type EmailValue = Pick<EmailNode, 'email'>;
+export type PhoneValue = Pick<PhoneNode, 'phone'>;
+
+type BasicDataSource = BasicDataValue & { id: ProfileData['id'] };
+type AdditionalInformationSource = Pick<ProfileData, 'id' | 'language'>;
+export type EditDataProfileSource =
+  | BasicDataSource
+  | AdditionalInformationSource
+  | AddressNode
+  | PhoneNode
+  | EmailNode;
+
+export type MultiItemProfileNode = AddressNode | PhoneNode | EmailNode;
+
+export const basicDataType = 'basic-data';
+export const additionalInformationType = 'additional-information';
+
+export type EditDataType =
+  | 'phones'
+  | 'emails'
+  | 'addresses'
+  | typeof basicDataType
+  | typeof additionalInformationType;
+
+export type EditDataValue =
+  | BasicDataValue
+  | AdditionalInformationValue
+  | AddressValue
+  | PhoneValue
+  | EmailValue;
+
+export type FormValues = {
+  firstName: string;
+  nickname: string;
+  lastName: string;
+  primaryEmail: PrimaryEmail;
+  primaryAddress: PrimaryAddress;
+  primaryPhone: PrimaryPhone;
+  language: Language;
+  addresses: AddressNode[];
+  emails: EmailNode[];
+  phones: PhoneNode[];
+};
+
+type SaveType = 'value' | 'set-primary' | 'remove' | undefined;
+
+export type EditData = {
+  readonly id: string;
+  readonly value: EditDataValue;
+  readonly primary?: boolean;
```

Comment:
This is only place where primary is defined.

"editData.value" only has value(s) needed in the input fields. 

Values are {phone:'<phone>'},  {email:'<email>'},  {address:'street', city...} etc.

Reply:
Right, the `EditDataValue`'s are those limited ones without the `primary` member. I don't quite get why there needs to be these limited types, why not just use the `AddressNode`, `EmailNode` and `PhoneNode` types directly (or include the `primary` member into the limited types). Perhaps I will learn it when the editing capabilities start appearing.

Reply:
Primary is not a value that user can write, so that's why it is not in the value-property. And checking changes in the value when user has changed the primary would cause false positives. And value is should be compatible with the form output.